File: leetcode-cn/15.三数之和.py
#
# @lc app=leetcode.cn id=15 lang=python3
#
# [15] 三数之和
#

# @lc code=start
from typing import List


class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        nums.sort()
        n = len(nums)
        ans = []

        for first in range(n - 2):
            if first > 0 and nums[first] == nums[first - 1]:
                continue

            second = first + 1
            third = n - 1

            if nums[first] > 0:
                continue

            while second < third:
                if nums[first] + nums[second] + nums[third] < 0:
                    second += 1
                elif nums[first] + nums[second] + nums[third] > 0:
                    third -= 1
                else:
                    ans.append([nums[first], nums[second], nums[third]])
                    while second < third and nums[second] == nums[second + 1]:
                        second += 1
                    second += 1

                    while second < third and nums[third] == nums[third - 1]:
                        third -= 1
                    third -= 1

        return ans


# @lc code=end

File: leetcode-cn/C++/15.三数之和.cpp
/*
 * @lc app=leetcode.cn id=15 lang=cpp
 *
 * [15] 三数之和
 */

// @lc code=start
#include <algorithm>
#include <iostream>
#include <vector>

using namespace std;
class Solution
{
public:
    vector<vector<int>> threeSum(vector<int> &nums)
    {
        int n = nums.size();
        sort(nums.begin(), nums.end());
        vector<vector<int>> ans;

        for (int i = 0; i < n; i++)
        {
            if (i > 0 && nums[i] == nums[i - 1])
            {
                continue;
            }

            int j = i + 1;
            int k = n - 1;

            while (j < k)
            {
                if (nums[i] + nums[j] + nums[k] < 0)
                {
                    j++;
                }
                else if (nums[i] + nums[j] + nums[k] > 0)
                {
                    k--;
                }
                else
                {
                    ans.push_back({nums[i], nums[j], nums[k]});
                    while (j < k && nums[j] == nums[j + 1])
                    {
                        j++;
                    }
                    j++;
                    while (j < k && nums[k] == nums[k - 1])
                    {
                        k--;
                    }
                    k--;
                }
            }
        }
        return ans;
    }
};
// @lc code=end
